include/flatbuffers/grpc.h

```diff
@@ -23,50 +23,271 @@
+                                              size_t new_size) override {
+    assert(old_p == GRPC_SLICE_START_PTR(slice_));
+    assert(old_size == GRPC_SLICE_LENGTH(slice_));
+    assert(new_size > old_size);
+    grpc_slice old_slice = slice_;
+    grpc_slice new_slice = grpc_slice_malloc(new_size);
+    uint8_t *new_p = GRPC_SLICE_START_PTR(new_slice);
+    memcpy(new_p + (new_size - old_size), old_p, old_size);
+    slice_ = new_slice;
+    grpc_slice_unref(old_slice);
+    return new_p;
+  }
+
+ private:
+  inline grpc_slice &get_slice(uint8_t *p, size_t size) {
+    assert(p == GRPC_SLICE_START_PTR(slice_));
+    assert(size == GRPC_SLICE_LENGTH(slice_));
+    return slice_;
+  }
+
+  grpc_slice slice_;
+
+  friend class MessageBuilder;
+};
+
+// SliceAllocatorMember is a hack to ensure that the MessageBuilder's
+// slice_allocator_ member is constructed before the FlatBufferBuilder, since
+// the allocator is used in the FlatBufferBuilder ctor.
+namespace detail {
+struct SliceAllocatorMember {
+  SliceAllocator slice_allocator_;
+};
+}
+
+// MessageBuilder is a gRPC-specific FlatBufferBuilder that uses SliceAllocator
+// to allocate gRPC buffers.
+class MessageBuilder : private detail::SliceAllocatorMember,
+                       public FlatBufferBuilder {
+ public:
+  explicit inline MessageBuilder(uoffset_t initial_size = 1024)
+    : FlatBufferBuilder(initial_size, &slice_allocator_, false),
+      initial_size_(initial_size) {}
+
+  MessageBuilder(const MessageBuilder &other) = delete;
+  MessageBuilder &operator=(const MessageBuilder &other) = delete;
+
+  inline ~MessageBuilder() {}
+
+  // GetMessage extracts the subslice of the buffer corresponding to the
+  // flatbuffers-encoded region and wraps it in a `Message<T>` to handle buffer
+  // ownership.
+  template <class T>
+  inline Message<T> GetMessage() {
+    vector_downward &vec = GetVectorDownward();
+    uint8_t *buf_data = vec.buf();     // pointer to memory
+    size_t buf_size = vec.capacity();  // size of memory
+    uint8_t *msg_data = vec.data();    // pointer to msg
+    size_t msg_size = vec.size();      // size of msg
+
+    // Do some sanity checks on data/size
+    assert(msg_data >= msg_data);
+    assert(msg_data + msg_size <= buf_data + buf_size);
+
+    // Calculate offsets from the buffer start
+    size_t begin = msg_data - buf_data;
+    size_t end = begin + msg_size;
+
+    // Get the slice we are working with (no refcount change)
+    grpc_slice slice = slice_allocator_.get_slice(buf_data, buf_size);
+
+    // Extract a subslice of the existing slice (increment refcount)
+    grpc_slice subslice = grpc_slice_sub(slice, begin, end);
+
+    // Wrap the subslice in a `Message<T>` (steal ref since incremented above)
+    Message<T> msg(subslice, STEAL_REF);
+
+    return msg;
+  }
+
+  template <class T>
+  inline Message<T> ReleaseMessage() {
+    Message<T> msg = GetMessage<T>();
+    Reset(initial_size_, &slice_allocator_, false);
```

Comment:
Like I said, even if `deallocate` is just an unref, there's still an allocation in `reset` that is not needed, and will just get deallocated shortly after if this `MessageBuilder` is abandoned.